# Perkenalan

Nama/Batch : Sandika/008

# Import Pustaka

In [ ]:
import pandas as pd
import numpy as np
import seaborn as sns
from sklearn import linear_model
from sklearn.linear_model import LinearRegression
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.metrics import r2_score
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score
from sklearn.preprocessing import PolynomialFeatures
from scipy import stats
from sklearn.preprocessing import MinMaxScaler
from sklearn.preprocessing import OrdinalEncoder
from scipy.stats import ttest_ind
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

# Soal 1

### Data Loading

In [ ]:
df_dummy= pd.read_csv('insurance.csv')
df_dummy.head()

age     sex     bmi  children smoker     region      charges
0   19  female  27.900         0    yes  southwest  16884.92400
1   18    male  33.770         1     no  southeast   1725.55230
2   28    male  33.000         3     no  southeast   4449.46200
3   33    male  22.705         0     no  northwest  21984.47061
4   32    male  28.880         0     no  northwest   3866.85520

In [ ]:
df= df_dummy.copy()

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1338 entries, 0 to 1337
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       1338 non-null   int64  
 1   sex       1338 non-null   object 
 2   bmi       1338 non-null   float64
 3   children  1338 non-null   int64  
 4   smoker    1338 non-null   object 
 5   region    1338 non-null   object 
 6   charges   1338 non-null   float64
dtypes: float64(2), int64(2), object(3)
memory usage: 73.3+ KB


Tidak ditemukan adanya data yang bernilai NaN

## EDA

In [ ]:
df.groupby('age').mean()['charges'].sort_values(ascending=False)

age
64    23275.530837
61    22024.457609
60    21979.418507
63    19884.998461
43    19267.278653
62    19163.856573
59    18895.869532
54    18758.546475
52    18256.269719
37    18019.911877
47    17653.999593
57    16447.185250
55    16164.545488
53    16020.930755
44    15859.396587
51    15682.255867
50    15663.003301
56    15025.515837
45    14830.199856
48    14632.500445
46    14342.590639
58    13878.928112
42    13061.038669
30    12719.110358
49    12696.006264
23    12419.820040
33    12351.532987
36    12204.476138
27    12184.701721
39    11778.242945
40    11772.251310
34    11613.528121
35    11307.182031
24    10648.015962
29    10430.158727
31    10196.980573
20    10159.697736
22    10012.932802
25     9838.365311
19     9747.909335
41     9653.745650
32     9220.300291
28     9069.187564
38     8102.733674
18     7086.217556
26     6133.825309
21     4730.464330
Name: charges, dtype: float64

mencari rata rata biaya asuransi berdasarkan umur

In [ ]:
df.describe()

age          bmi     children       charges
count  1338.000000  1338.000000  1338.000000   1338.000000
mean     39.207025    30.663397     1.094918  13270.422265
std      14.049960     6.098187     1.205493  12110.011237
min      18.000000    15.960000     0.000000   1121.873900
25%      27.000000    26.296250     0.000000   4740.287150
50%      39.000000    30.400000     1.000000   9382.033000
75%      51.000000    34.693750     2.000000  16639.912515
max      64.000000    53.130000     5.000000  63770.428010

In [ ]:
df.corr().sort_values('charges', ascending=False, key=abs).index

Index(['charges', 'age', 'bmi', 'children'], dtype='object')

mencari kolom dengan korelasi tertinggi dengan charges

## Data Preprocessing

In [ ]:
data_inf= df.sample(10, random_state=21) 
data_inf

age     sex     bmi  children smoker     region      charges
831    36  female  25.840         0     no  northwest   5266.36560
1287   36  female  25.900         1     no  southwest   5472.44900
1259   52  female  23.180         0     no  northeast  10197.77220
827    36    male  28.025         1    yes  northeast  20773.62775
933    45  female  35.300         0     no  southwest   7348.14200
779    53    male  28.880         0     no  northwest   9869.81020
854    49  female  23.845         3    yes  northeast  24106.91255
1020   51    male  37.000         0     no  southwest   8798.59300
1037   45  female  30.495         1    yes  northwest  39725.51805
1096   51  female  34.960         2    yes  northeast  44641.19740

Proses pengambilan data inference

In [ ]:
data_train_test = df.drop(data_inf.index)
data_train_test

age     sex     bmi  children smoker     region      charges
0      19  female  27.900         0    yes  southwest  16884.92400
1      18    male  33.770         1     no  southeast   1725.55230
2      28    male  33.000         3     no  southeast   4449.46200
3      33    male  22.705         0     no  northwest  21984.47061
4      32    male  28.880         0     no  northwest   3866.85520
...   ...     ...     ...       ...    ...        ...          ...
1333   50    male  30.970         3     no  northwest  10600.54830
1334   18  female  31.920         0     no  northeast   2205.98080
1335   18  female  36.850         0     no  southeast   1629.83350
1336   21  female  25.800         0     no  southwest   2007.94500
1337   61  female  29.070         0    yes  northwest  29141.36030

[1328 rows x 7 columns]

proses pengambilan data train_test

In [ ]:
data_train_test.reset_index(drop=True,inplace=True)
data_inf.reset_index(drop=True,inplace=True)

proses mereset index dari data inf dan train test

In [ ]:
X= data_train_test.drop(['charges'], axis=1)
y= data_train_test['charges']
X

age     sex     bmi  children smoker     region
0      19  female  27.900         0    yes  southwest
1      18    male  33.770         1     no  southeast
2      28    male  33.000         3     no  southeast
3      33    male  22.705         0     no  northwest
4      32    male  28.880         0     no  northwest
...   ...     ...     ...       ...    ...        ...
1323   50    male  30.970         3     no  northwest
1324   18  female  31.920         0     no  northeast
1325   18  female  36.850         0     no  southeast
1326   21  female  25.800         0     no  southwest
1327   61  female  29.070         0    yes  northwest

[1328 rows x 6 columns]

In [ ]:
X_train, X_test, y_train, y_test=train_test_split(X, y, test_size=0.20, random_state=21) #test_size persentase data yang diambil 0.20= 20%

print(X_train.shape)
print(X_test.shape)

(1062, 6)
(266, 6)


In [ ]:
num_columns = X_train.select_dtypes(include=np.number).columns.tolist()
cat_columns = X_train.select_dtypes(include=['object']).columns.tolist()
print('Numeric Columns:', num_columns)
print('categorical Columns:', cat_columns)

Numeric Columns: ['age', 'bmi', 'children']
categorical Columns: ['sex', 'smoker', 'region']


In [ ]:
X_train_num = X_train[num_columns]
X_train_cat = X_train[cat_columns]

X_test_num = X_test[num_columns]
X_test_cat = X_test[cat_columns]

In [ ]:
scaler = MinMaxScaler ()
X_train_num_scaled = scaler.fit_transform(X_train_num)
X_test_num_scaled= scaler.transform(X_test_num)

In [ ]:
encoder =OrdinalEncoder()
X_train_cat_encoded = encoder.fit_transform(X_train_cat)
X_test_cat_encoded=encoder.transform(X_test_cat)
X_train_cat_encoded

array([[0., 0., 1.],
       [1., 0., 0.],
       [0., 0., 0.],
       ...,
       [0., 0., 0.],
       [0., 0., 2.],
       [1., 0., 2.]])

proses encode dan mengskalasi kolom number dan juga kategori

In [ ]:
X_train_final = np.concatenate([X_train_num_scaled, X_train_cat_encoded], axis=1)
X_test_final = np.concatenate([X_test_num_scaled, X_test_cat_encoded], axis=1)

In [ ]:
X_train_final_df=pd.DataFrame(X_train_final,columns=[num_columns+cat_columns])
X_train_final_df

age       bmi children  sex smoker region
0     0.934783  0.395016      0.0  0.0    0.0    1.0
1     0.000000  0.238056      0.0  1.0    0.0    0.0
2     1.000000  0.619992      0.0  0.0    0.0    0.0
3     0.391304  0.146496      0.6  0.0    0.0    0.0
4     0.347826  0.273028      0.2  0.0    0.0    2.0
...        ...       ...      ...  ...    ...    ...
1057  0.565217  0.418560      0.4  1.0    0.0    1.0
1058  0.913043  0.212447      0.0  0.0    0.0    2.0
1059  0.565217  0.541512      0.0  0.0    0.0    0.0
1060  0.152174  0.697095      0.2  0.0    0.0    2.0
1061  0.173913  0.512323      0.0  1.0    0.0    2.0

[1062 rows x 6 columns]

didapat data frame baru yang sudah diencode dan discale

## Model Definition

In [ ]:
model_lin_reg =LinearRegression()
model = LinearRegression(fit_intercept=True)

## Model Training

In [ ]:
model_lin_reg.fit(X_train_final, y_train)

LinearRegression()

## Model Evaluation

In [ ]:
metric = pd.DataFrame()
metric['models'] = ['OLR','Poly3','OLR_reg','Poly3_reg']
r2_train = []
r2_test = []

In [ ]:
model.fit(X_train_final,y_train)

LinearRegression()

In [ ]:
print('-------- TRAIN ---------')
y_pred = model.predict(X_train_final)
print('MAE:',mean_absolute_error(y_train,y_pred))
print('MSE:',mean_squared_error(y_train,y_pred))
print('RMSE:',mean_squared_error(y_train,y_pred,squared=False))
print('R^2 score:',r2_score(y_train,y_pred))
r2_train.append(r2_score(y_train,y_pred))

print('-------- TEST ---------')
y_pred = model.predict(X_test_final)
print('MAE:',mean_absolute_error(y_test,y_pred))
print('MSE:',mean_squared_error(y_test,y_pred))
print('RMSE:',mean_squared_error(y_test,y_pred,squared=False))
print('R^2 score:',r2_score(y_test,y_pred))
r2_test.append(r2_score(y_test,y_pred))

-------- TRAIN ---------
MAE: 4194.148134905426
MSE: 36828525.249950044
RMSE: 6068.651023905563
R^2 score: 0.7484326996071404
-------- TEST ---------
MAE: 4173.406894588773
MSE: 35724787.9861652
RMSE: 5977.021665191218
R^2 score: 0.7531382006173861


In [ ]:
print('Intecept : ' ,model_lin_reg.intercept_)
print('Koefisien : ', model_lin_reg.coef_)

Intecept :  -1730.2165699561701
Koefisien :  [11747.60510624 12678.13483677  2811.68565648  -148.35636534
 23820.68686742  -472.27823183]


In [ ]:
X_train_final_df.columns

MultiIndex([(     'age',),
            (     'bmi',),
            ('children',),
            (     'sex',),
            (  'smoker',),
            (  'region',)],
           )

## Model Inference

In [ ]:
data_inf

age     sex     bmi  children smoker     region      charges
0   36  female  25.840         0     no  northwest   5266.36560
1   36  female  25.900         1     no  southwest   5472.44900
2   52  female  23.180         0     no  northeast  10197.77220
3   36    male  28.025         1    yes  northeast  20773.62775
4   45  female  35.300         0     no  southwest   7348.14200
5   53    male  28.880         0     no  northwest   9869.81020
6   49  female  23.845         3    yes  northeast  24106.91255
7   51    male  37.000         0     no  southwest   8798.59300
8   45  female  30.495         1    yes  northwest  39725.51805
9   51  female  34.960         2    yes  northeast  44641.19740

In [ ]:
data_inf_num = data_inf[num_columns]
data_inf_cat = data_inf[cat_columns]

In [ ]:
data_inf_num_scaled = scaler.transform(data_inf_num)
data_inf_cat_encoded = encoder.transform(data_inf_cat)

In [ ]:
data_inf_final= np.concatenate([data_inf_num_scaled, data_inf_cat_encoded], axis=1)
data_inf_final

array([[0.39130435, 0.2485199 , 0.        , 0.        , 0.        ,
        1.        ],
       [0.39130435, 0.25017211, 0.2       , 0.        , 0.        ,
        3.        ],
       [0.73913043, 0.17527193, 0.        , 0.        , 0.        ,
        0.        ],
       [0.39130435, 0.30868787, 0.2       , 1.        , 1.        ,
        0.        ],
       [0.58695652, 0.50901831, 0.        , 0.        , 0.        ,
        3.        ],
       [0.76086957, 0.33223186, 0.        , 1.        , 0.        ,
        1.        ],
       [0.67391304, 0.19358392, 0.6       , 0.        , 1.        ,
        0.        ],
       [0.7173913 , 0.55583092, 0.        , 1.        , 0.        ,
        3.        ],
       [0.58695652, 0.37670384, 0.2       , 0.        , 1.        ,
        1.        ],
       [0.7173913 , 0.49965579, 0.4       , 0.        , 1.        ,
        0.        ]])

In [ ]:
y_pred_inf= model_lin_reg.predict(data_inf_final)
y_pred_inf = pd.DataFrame(y_pred_inf, columns=['charges-Prediction'])
y_pred_inf

charges-Prediction
0         5545.162896
1         5183.890502
2         9174.917013
3        31014.926457
4        10201.684960
5        10799.624342
6        34148.629023
7        12179.121519
8        33851.764723
9        37977.477783

In [ ]:
pd.concat([data_inf, y_pred_inf], axis=1)

age     sex     bmi  children smoker     region      charges  \
0   36  female  25.840         0     no  northwest   5266.36560   
1   36  female  25.900         1     no  southwest   5472.44900   
2   52  female  23.180         0     no  northeast  10197.77220   
3   36    male  28.025         1    yes  northeast  20773.62775   
4   45  female  35.300         0     no  southwest   7348.14200   
5   53    male  28.880         0     no  northwest   9869.81020   
6   49  female  23.845         3    yes  northeast  24106.91255   
7   51    male  37.000         0     no  southwest   8798.59300   
8   45  female  30.495         1    yes  northwest  39725.51805   
9   51  female  34.960         2    yes  northeast  44641.19740   

   charges-Prediction  
0         5545.162896  
1         5183.890502  
2         9174.917013  
3        31014.926457  
4        10201.684960  
5        10799.624342  
6        34148.629023  
7        12179.121519  
8        33851.764723  
9        37977.477783

proses penggabungan data inference di kedalam data_inf

## Kesimpulan

In [ ]:
print('Intecept : ' ,model_lin_reg.intercept_)
print('Koefisien : ', model_lin_reg.coef_)

Intecept :  -1730.2165699561701
Koefisien :  [11747.60510624 12678.13483677  2811.68565648  -148.35636534
 23820.68686742  -472.27823183]


In [ ]:
X_train_final_df.columns

MultiIndex([(     'age',),
            (     'bmi',),
            ('children',),
            (     'sex',),
            (  'smoker',),
            (  'region',)],
           )

- Persamaan regresi :

charges= -1730.2165699561701 + 11747.60510624age + 12678.13483677bmi + 2811.68565648children - 148.35636534sex +  23820.68686742smoker - 472.27823183region

- Parameter yang paling berpengaruh terhadap kolom 'charges' adalah kolom 'smoker' sementara parameter yang paling tidak berpengaruh adalah kolom 'sex'

Nilai evaluasi model di data train dan test:
- TRAIN 
- MAE: 4194.148134905426
- MSE: 36828525.249950044
- RMSE: 6068.651023905563
- R^2 score: 0.7484326996071404
- TEST
- MAE: 4173.406894588773
- MSE: 35724787.9861652
- RMSE: 5977.021665191218
- R^2 score: 0.7531382006173861

# Soal 2

## Data Loading

In [ ]:
data_log = pd.read_csv('https://raw.githubusercontent.com/fahmimnalfrzki/Dataset/main/GalaxyMorphology.csv')
data_log = data_log[data_log['CNN2classes1stClass']!='Converted']
data_log

Unnamed: 0     TType          K            C         A         S  \
0               0  1.854829   6.560199     0.357051  0.932151  0.850992   
1               1  4.815538  25.950829     0.313292  0.610999  0.723122   
2               2  0.678056   6.867416     0.408286  0.881524  0.813231   
3               3 -0.547708  11.877941 -9999.999999  0.838779  0.816965   
4               4  2.028543  20.458965     0.446453  0.839216  0.801140   
...           ...       ...        ...          ...       ...       ...   
19995       19995  4.631655  29.018608     0.358020  0.602471  0.832848   
19996       19996 -2.179893  15.245582     0.470026  0.842624  0.868289   
19997       19997  1.876151   6.907812     0.323334  0.830340  0.950820   
19998       19998  5.053256  27.981613     0.393574  0.632766  0.825990   
19999       19999 -0.558501   6.557439     0.346322  0.878231  0.882251   

             G2         H  CNN2classes1stClass  
0      1.588183  0.697214                    1  
1      1.533653  0.719412                    1  
2      1.743747  0.750872                    1  
3      1.672529  0.771995                    1  
4      1.107665  0.610453                    1  
...         ...       ...                  ...  
19995  1.611594  0.735315                    1  
19996  1.313089  0.649996                    1  
19997  1.406042  0.671217                    1  
19998  1.659363  0.379448                    1  
19999  1.616493  0.738060                    1  

[20000 rows x 9 columns]

In [ ]:
data_log.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 20000 entries, 0 to 19999
Data columns (total 9 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Unnamed: 0           20000 non-null  int64  
 1   TType                20000 non-null  float64
 2   K                    20000 non-null  float64
 3   C                    20000 non-null  float64
 4   A                    20000 non-null  float64
 5   S                    20000 non-null  float64
 6   G2                   20000 non-null  float64
 7   H                    20000 non-null  float64
 8   CNN2classes1stClass  20000 non-null  int64  
dtypes: float64(7), int64(2)
memory usage: 1.5 MB


tidak ada data bernilai NaN

## EDA

In [ ]:
kolom=['C', 'A', 'S','G2','H', 'CNN2classes1stClass']
new_data_log= data_log[kolom]


proses pengambilan kolom untuk digunakan ke dalam logistic regression

In [ ]:
new_data_log

C         A         S        G2         H  \
0         0.357051  0.932151  0.850992  1.588183  0.697214   
1         0.313292  0.610999  0.723122  1.533653  0.719412   
2         0.408286  0.881524  0.813231  1.743747  0.750872   
3     -9999.999999  0.838779  0.816965  1.672529  0.771995   
4         0.446453  0.839216  0.801140  1.107665  0.610453   
...            ...       ...       ...       ...       ...   
19995     0.358020  0.602471  0.832848  1.611594  0.735315   
19996     0.470026  0.842624  0.868289  1.313089  0.649996   
19997     0.323334  0.830340  0.950820  1.406042  0.671217   
19998     0.393574  0.632766  0.825990  1.659363  0.379448   
19999     0.346322  0.878231  0.882251  1.616493  0.738060   

       CNN2classes1stClass  
0                        1  
1                        1  
2                        1  
3                        1  
4                        1  
...                    ...  
19995                    1  
19996                    1  
19997                    1  
19998                    1  
19999                    1  

[20000 rows x 6 columns]

semua data merupakan data numerik


In [ ]:
y = new_data_log['CNN2classes1stClass']

In [ ]:
#H0: Tidak ada perbedaan yang signifikan galaksi elips dan galaksi spiral terhadap suatu kolom
#H1: Ada perbedaan yang signifikan antara galaksi elips dan galaksi spiral terhadap suatu kolom
lab1_id = np.where(y==1)[0]
lab0_id = np.where(y==0)[0]

t,p_val = ttest_ind(new_data_log.iloc[lab1_id],new_data_log.iloc[lab0_id])

p_val_num = pd.DataFrame({'COLS':new_data_log.columns,'P-Value':p_val})
drop_cols = p_val_num[p_val_num['P-Value']>0.05]
drop_cols

COLS   P-Value
1    A  0.574941
2    S  0.168806
3   G2  0.796503

In [ ]:
# kolom A S dan G2 memenuhi H0
print("\U0001f600")

😀


In [ ]:
# Data set baru yang akan digunakan
data_new = new_data_log.drop(columns=drop_cols['COLS'])
data_new.head()

C         H  CNN2classes1stClass
0     0.357051  0.697214                    1
1     0.313292  0.719412                    1
2     0.408286  0.750872                    1
3 -9999.999999  0.771995                    1
4     0.446453  0.610453                    1

In [ ]:
data_new.isnull().sum()

C                      0
H                      0
CNN2classes1stClass    0
dtype: int64

In [ ]:
data_new.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 20000 entries, 0 to 19999
Data columns (total 3 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   C                    20000 non-null  float64
 1   H                    20000 non-null  float64
 2   CNN2classes1stClass  20000 non-null  int64  
dtypes: float64(2), int64(1)
memory usage: 625.0 KB


## Data Preprocessing

In [ ]:
X2= data_new.drop(['CNN2classes1stClass'], axis=1)
y2= data_new['CNN2classes1stClass'].replace({1:'galaksi spiral',0:'galaksi elips'})
X2

C         H
0         0.357051  0.697214
1         0.313292  0.719412
2         0.408286  0.750872
3     -9999.999999  0.771995
4         0.446453  0.610453
...            ...       ...
19995     0.358020  0.735315
19996     0.470026  0.649996
19997     0.323334  0.671217
19998     0.393574  0.379448
19999     0.346322  0.738060

[20000 rows x 2 columns]

In [ ]:
y2.head()

0    galaksi spiral
1    galaksi spiral
2    galaksi spiral
3    galaksi spiral
4    galaksi spiral
Name: CNN2classes1stClass, dtype: object

In [ ]:
X_train2,X_test2,y_train2,y_test2 = train_test_split(X2,y2,train_size=0.8,stratify=y2)
X_train_scaled2 = scaler.fit_transform(X_train2)
X_test_scaled2 = scaler.transform(X_test2)

## Model Definition

In [ ]:
model = LogisticRegression(penalty='l2',C=0.01,solver='newton-cg')

model.fit(X_train_scaled2,y_train2)

LogisticRegression(C=0.01, solver='newton-cg')

## Model Evaluation

In [ ]:
y_pred_train2 = model.predict(X_train_scaled2)
y_pred_test2 = model.predict(X_test_scaled2)
print('Accuracy Train:',accuracy_score(y_train2,y_pred_train2))
print('Accuracy Test:',accuracy_score(y_test2,y_pred_test2))

Accuracy Train: 0.8070625
Accuracy Test: 0.80725


In [ ]:
print('Score Train:',model.score(X_train_scaled2, y_train2))
print('Score Test:',model.score(X_test_scaled2, y_test2))

Score Train: 0.8070625
Score Test: 0.80725


proses evaluasi model

## Model Inference

In [ ]:
data_sample = X2.sample(2)
data_sample

C         H
9010  0.379552  0.396207
3065  0.350852  0.685838

In [ ]:
data_inference = scaler.transform(data_sample)
kategori= model.predict(data_inference)

print('Label:', (kategori))

Label: [1 1]


In [ ]:
lk = [] 
j= 0 
for i in data_new['CNN2classes1stClass']:
    if i==1:
        lk.append('galaksi spiral')
    else:
        lk.append('galaksi elips')
    j=j+1 

data_new['Kategori galaxi'] = lk 

In [ ]:
data_new.drop(['Label'], axis = 1)

C         H  CNN2classes1stClass Kategori galaxi
0         0.357051  0.697214                    1  galaksi spiral
1         0.313292  0.719412                    1  galaksi spiral
2         0.408286  0.750872                    1  galaksi spiral
3     -9999.999999  0.771995                    1  galaksi spiral
4         0.446453  0.610453                    1  galaksi spiral
...            ...       ...                  ...             ...
19995     0.358020  0.735315                    1  galaksi spiral
19996     0.470026  0.649996                    1  galaksi spiral
19997     0.323334  0.671217                    1  galaksi spiral
19998     0.393574  0.379448                    1  galaksi spiral
19999     0.346322  0.738060                    1  galaksi spiral

[20000 rows x 4 columns]

## Kesimpulan

- Nilai akurasi model logistic regression adalah 80.7 %
- kolom A S dan G2 tidak digunakan dalam logistic regression karena memenuhi H0 dimana nilai p-value> 0.05 dimana H0 adalah:
- H0: Tidak ada perbedaan yang signifikan galaksi elips dan galaksi spiral terhadap suatu kolom